beacon-chain/core/balances/rewards_penalties.go

```diff
@@ -196,3 +196,194 @@ func InclusionDistRewards(
+func InactivityHeadPenalties(
+	state *pb.BeaconState,
+	headAttesterIndices []uint32,
+	totalBalance uint64) *pb.BeaconState {
+
+	baseRewardQuotient := baseRewardQuotient(totalBalance)
+	allValidatorIndices := validators.AllActiveValidatorsIndices(state)
+	didNotAttestIndices := slices.Not(headAttesterIndices, allValidatorIndices)
+
+	for _, index := range didNotAttestIndices {
+		state.ValidatorBalances[index] -=
+			baseReward(state, index, baseRewardQuotient)
+	}
+	return state
+}
+
+// InactivityExitedPenalties applies additional (2x) penalties
+// to inactive validators with status EXITED_WITH_PENALTY.
+//
+// Spec pseudocode definition:
+//    Any validator index with status == EXITED_WITH_PENALTY,
+//    loses 2 * inactivity_penalty(state, index, epochs_since_finality) +
+//    base_reward(state, index)
+func InactivityExitedPenalties(
+	state *pb.BeaconState,
+	totalBalance uint64,
+	epochsSinceFinality uint64) *pb.BeaconState {
+
+	baseRewardQuotient := baseRewardQuotient(totalBalance)
+	allValidatorIndices := validators.AllValidatorsIndices(state)
+
+	for _, index := range allValidatorIndices {
+		if state.ValidatorRegistry[index].Status == pb.ValidatorRecord_EXITED_WITH_PENALTY {
+			state.ValidatorBalances[index] -=
+				2*inactivityPenalty(state, index, baseRewardQuotient, epochsSinceFinality) +
+					baseReward(state, index, baseRewardQuotient)
+		}
+	}
+	return state
+}
+
+// InactivityInclusionPenalties applies penalties in relation with
+// inclusion delay to inactive validators.
+//
+// Spec pseudocode definition:
+//    Any validator index in previous_epoch_attester_indices loses
+//    base_reward(state, index) - base_reward(state, index) *
+//    MIN_ATTESTATION_INCLUSION_DELAY // inclusion_distance(state, index)
+func InactivityInclusionPenalties(
+	state *pb.BeaconState,
+	attesterIndices []uint32,
+	totalBalance uint64) (*pb.BeaconState, error) {
+
+	baseRewardQuotient := baseRewardQuotient(totalBalance)
+
+	for _, index := range attesterIndices {
+		inclusionDistance, err := epoch.InclusionDistance(state, index)
+		if err != nil {
+			return nil, fmt.Errorf("could not get inclusion distance: %v", err)
+		}
+		baseReward := baseReward(state, index, baseRewardQuotient)
+		state.ValidatorBalances[index] -= baseReward -
+			baseReward*params.BeaconConfig().MinAttestationInclusionDelay/
+				inclusionDistance
+	}
+	return state, nil
+}
+
+// AttestationInclusionRewards awards the the beacon
+// proposers who included previous epoch attestations.
+//
+// Spec pseudocode definition:
+//    For each index in previous_epoch_attester_indices,
+//    we determine the proposer proposer_index =
+//    get_beacon_proposer_index(state, inclusion_slot(state, index))
+//    and set state.validator_balances[proposer_index] +=
+//    base_reward(state, index) // INCLUDER_REWARD_QUOTIENT
+func AttestationInclusionRewards(
+	state *pb.BeaconState,
+	totalBalance uint64,
+	attesterIndices []uint32) (*pb.BeaconState, error) {
```

Comment:
can we change this parameter to be more clear it is the previous epoch's attester indices? prevEpochAttesterIndices